src/main/java/com/google/devtools/build/lib/bazel/bzlmod/SingleExtensionEvalFunction.java

```diff
@@ -258,43 +255,129 @@ private SingleExtensionEvalValue tryGettingValueFromLockFile(
+        RunnableExtension extension,
+        LockFileModuleExtension lockedExtension,
+        ImmutableMap<ModuleKey, ModuleExtensionUsage> extensionUsages,
+        ImmutableMap<ModuleKey, ModuleExtensionUsage> lockedExtensionUsages)
+        throws InterruptedException, NeedsSkyframeRestartException {
+      try {
+        // Put faster diff detections earlier, so that we can short-circuit in UPDATE mode.
+        if (!Arrays.equals(extension.getBzlTransitiveDigest(),
+            lockedExtension.getBzlTransitiveDigest())) {
+          reportDiff(
+              "The implementation of the extension '"
+                  + extensionId
+                  + "' or one of its transitive .bzl files has changed");
+        }
+        if (!extension.getEnvVars().equals(lockedExtension.getEnvVariables())) {
+          reportDiff(
+              "The environment variables the extension '"
+                  + extensionId
+                  + "' depends on (or their values) have changed");
+        }
+        // Check extension data in lockfile is still valid, disregarding usage information that is not
+        // relevant for the evaluation of the extension.
+        if (!ModuleExtensionUsage.trimForEvaluation(extensionUsages)
+            .equals(ModuleExtensionUsage.trimForEvaluation(lockedExtensionUsages))) {
+          reportDiff("The usages of the extension '" + extensionId + "' have changed");
+        }
+        if (didRepoMappingsChange(env, lockedExtension.getRecordedRepoMappingEntries())) {
+          reportDiff("The repo mapping of certain repos used by the extension '" + extensionId
+              + "' have changed");
+        }
+        if (didFilesChange(env, lockedExtension.getAccumulatedFileDigests())) {
+          reportDiff(
+              "One or more files the extension '" + extensionId + "' is using have changed");
+        }
+      } catch (DiffFoundEarlyExitException ignored) {
+        // ignored
+      }
+    }
+
+    public boolean anyDiffsDetected() {
+      return diffDetected;
+    }
+
+    public String getRecordedDiffMessages() {
+      return String.join(",", diffMessages.build());
+    }
+  }
+
+  private static boolean didRepoMappingsChange(Environment env,
+      ImmutableTable<RepositoryName, String, RepositoryName> recordedRepoMappings)
+      throws InterruptedException, NeedsSkyframeRestartException {
+    SkyframeLookupResult result = env.getValuesAndExceptions(
+        recordedRepoMappings.rowKeySet().stream().map(
+            RepositoryMappingValue::key).collect(toImmutableSet()));
+    if (env.valuesMissing()) {
+      // This shouldn't really happen, since the RepositoryMappingValues of any recorded repos
+      // should have already been requested by the time we load the .bzl for the extension. And this
+      // method is only called if the transitive .bzl digest hasn't changed.
+      // However, we pretend to restart Skyframe anyway because we're good citizens.
```

Comment:
Nit: this isn't "pretend", so just say "we restart Skyframe anyway because ..."

Reply:
reworded.